### ISSUE TITLE
Exception Class for Connection Errors 

### ISSUE THREAD

**Issue by tf** [pushpad/pushpad-ruby#7]
When communicating with Pushpad from background jobs, sometimes there are errors when connection problems occur or when the site is under maintenance etc. Since these errors are of a temporary nature, I'd like to have an easy way to ignore them in exception notifications or add some special logging.

Since `pushpad-ruby` currently simply uses `Net:HTTP`, there is a rather wide range of possible errors I have to look out for. So far I've seen `Errno::ECONNRESET` and `Net::OpenTimeout`, but looking at [Faradays `Net::Http` adatper](https://github.com/lostisland/faraday/blob/master/lib/faraday/adapter/net_http.rb#L12) it seems like there could potentially be a whole lot more.

Then there are [more specific errors like `Pushpad::DeliveryError`](https://github.com/pushpad/pushpad-ruby/blob/master/lib/pushpad/notification.rb#L97) which normally point to domain level errors (i.e. failed validation) but at the moment also come up when Pushpad responds with 503, which is also rather one of those recoverable transport level errors. 

Ideally there would be an exception base class like `Pushpad::ConnectionError` that encapsulates all of the different `Net::HTTP` errors. This would make handling these errors very convenient.

In addition it might make sense to add more exception classes for specific HTTP response codes (422, 404, 403, 503).

For the first part I see two options: 

* Rewriting `Pushpad::Request` to use something like Faraday, relying on its [unification of connection errors](https://github.com/lostisland/faraday/blob/master/lib/faraday/adapter/net_http.rb#L43) and re-wrapping its exceptions to decouple the public interface of this library from the HTTP client library choice.

* Port some of Faraday's exception handling code if the goal is to keep having no gem dependencies. For a lightweight library this one, that might be a selling point.

**Comment by collimarco** [comment]
Thank you for your feedback. I understand your point, but I think it needs some more thinking.

What would you like to achieve by catching the exceptions? What about a solution like the following?

Connection errors -> log and retry
422 -> validation error: permanent failure, just log
503 and other status codes -> log and retry

Is the above example reasonable? If so, you can just "log and retry" for all kinds of exceptions. Indeed 422 has no side effects if retried multiple times (and you will eventually fix your application).

Basically, **if you use Sidekiq (or other workers) I would let them manage the exceptions entirely, without any rescue in your application code**.

**Comment by tf** [comment]
Yes, this is basically what I am doing. A couple of thoughts:

- I do not want to let all of these exceptions bubble up to the job framework (Resque in my case), because these failed jobs queue up in the "failed jobs" list and make other more severe errors harder to find, when there is actually nothing wrong.

- I am thus rescuing all 14 `Net::HTTP` exceptions. This works, but I think it would be a nicer API if `pushpad-ruby` had its own single exception class for these cases. The original exception would automatically be available via `Exception#cause`. To me the fact that this library uses `Net::HTTP` is an implementation detail.

- Telling apart 422 from 503 currently requires parsing the text in `DeliveryError#message`. It would be much nicer if this information was available in a machine readable format. Using different exception classes looks like the most idiomatic way to go here.

**Comment by collimarco** [comment]
> I do not want to let all of these exceptions bubble up to the job framework (Resque in my case), because these failed jobs queue up in the "failed jobs" list and make other more severe errors harder to find, when there is actually nothing wrong. I am thus rescuing all 14 Net::HTTP exceptions [...]

Does that mean that you don't retry if there's a Net::HTTP exception? You should retry if there are connection errors. Maybe you can limit the number of retries - I don't know much about Resque but you can do it in Sidekiq for example.

> To me the fact that this library uses Net::HTTP is an implementation detail.
Yes, it is an implementation detail, but it's also true that Net::HTTP is part of Ruby (it's in the standard library, it's not a gem).

In any case, we may consider to catch all the Net::HTTP exceptions (or [maybe all the `StandardError`s](https://stackoverflow.com/a/11802674/51387)) inside [Pushpad::Request#perform](https://github.com/pushpad/pushpad-ruby/blob/master/lib/pushpad/request.rb#L24) and raise a `Pushpad::IOException`. However I still have some doubts about this point:
- isn't it the same if you rescue `StandardError` in your application? e.g.
```
begin
    notification.broadcast
rescue => e
    # ...
end
```
- should we raise a new `Pushpad::IOException` or use the existing `Notification::DeliveryError`? (and set a status code attribute to nil or zero, see below)

We can also add an attribute (`status_code`) to the `DeliveryError` to make it easier to get the HTTP status code. (I have no doubts about this second point).
https://github.com/pushpad/pushpad-ruby/blob/master/lib/pushpad/notification.rb#L101
https://github.com/pushpad/pushpad-ruby/blob/master/lib/pushpad/notification.rb#L6

**Comment by collimarco** [comment]
In short, if we add a `DeliveryError#status_code` you should be able to manage all exceptions with:

```ruby
begin
    notification.broadcast
rescue Pushpad::Notification::DeliveryError => e
    case e.status_code # currently not implemented (as a workaround you can get it with a regex from the exception message)
    when 422
        # ...
    when 503
        # ...
    end
rescue => e
    # ...
end
```

**Comment by tf** [comment]
> Does that mean that you don't retry if there's a Net::HTTP exception?

The jobs are scheduled to run every few minutes anyway to fetch notification delivery statistics. If a job fails it will automatically be retried next time it is scheduled.

> if we add a DeliveryError#status_code you should be able to manage all exceptions

Adding the status code field to the error would be fine. Based on that attribute one could even consider adding a `recoverable?` method, which indicates whether the request should be retried. That would help contain the knowledge of possible status codes inside that class.

> isn't it the same if you rescue StandardError in your application

Rescuing `StandardError` in my application would also hide all other exceptions that Pushpad or my code interacting with Pushpad might raise. 

Re-raising `StandardError` from `Pushpad::Request#perform` as `Pushpad::IOException` might be ok. We would have to be careful not to swallow other user facing errors like from [the auth token validation](https://github.com/pushpad/pushpad-ruby/blob/master/lib/pushpad/request.rb#L48).

Also I'm not sure if replacing the low level exceptions with one `IOException` might be a breaking change (even it is one that I'd like to see). Then again the gem is still 0.x

**Comment by collimarco** [comment]
> Rescuing StandardError in my application would also hide all other exceptions that Pushpad or my code interacting with Pushpad might raise.

There aren't many exceptions that can be raised if you just wrap `#broadcast` or `deliver_to` calls.

> Also I'm not sure if replacing the low level exceptions with one IOException might be a breaking change (even it is one that I'd like to see). Then again the gem is still 0.x

Yes, that's a point I was considering too: I prefer to avoid breaking changes if possible.

What I don't like about adding a new `Pushpad::IOException` is that, like in other libraries, you end up with code like:

```
begin
  ..
rescue IOException
  ..
rescue => e
  # there might be other exceptions... let's add a general clause anyway
end
```

Unless there's a real issue, something that you can't overcome without this feature, I prefer to wait. Let's see if other people requests that and what are the use cases.

At the moment, if you need a `DeliveryError#status_code`, feel free to make a pull request _for that_ and I will accept it.

**Comment by tf** [comment]
I'm also totally fine with waiting. If I feel the need, I'll create a PR for `status_code`.

Thanks.
